### PR TITLE
Implement image level QC

### PR DESCRIPTION
**Pull request by constantinpape**

### CODE REVIEW

**Comment by constantinpape** [comment]
@imagirom It would be great if you can have a look at the changes in cell_level_analysis, especially at this:
https://github.com/hci-unihd/batchlib/blob/image-qc/batchlib/analysis/cell_level_analysis.py#L818-L819

File: batchlib/analysis/cell_analysis_qc.py [hunk 1]
```diff
@@ -169,13 +208,135 @@ def run(self, input_files, output_files):
...
+
+        # check for negative ratios
+        if self.outlier_criteria['check_ratios']:
+            ratios = compute_ratios(control_stats, infected_stats, serum_key=self.serum_key)
+            for name, val in ratios.items():
+                if not name.startswith('ratio'):
+                    continue
+                if val < 0.:
+                    is_outlier = 1
+                    outlier_type += f'{name} is negative'
```
Comment (imagirom) [inline batchlib/analysis/cell_analysis_qc.py:292]: You only do this for wells, correct? We have yet to add the image-level outlier detection to the well- and plate-wise background calculation. If this is only done for wells we can first calculate the feature tables per image, then detect the image outliers, then compute the bg values (ignoring outlier images) and only then do the well-level outlier control. If we also want to do this on an image level, it gets more complicated (But I do not see the need for it)

File: batchlib/analysis/cell_level_analysis.py [hunk 2]
```diff
@@ -685,20 +710,31 @@ def load_image_outliers(self, input_files):
                         for ii in range(len(table))}
         return outlier_dict
 
-    def load_cell_outliers(self, input_file):
-        with open_file(input_file, 'r') as f:
-            if not self.has_table(f, self.cell_outlier_table):
-                logger.warning(f"{self.name}: load_cell_outliers: did not find a cell outlier table")
```
Comment (imagirom) [inline batchlib/analysis/cell_level_analysis.py:691]: Why are we getting rid of this? Or did this just move somewhere else?

File: batchlib/analysis/cell_level_analysis.py [hunk 3]
```diff
@@ -767,11 +802,25 @@ def write_well_table(self, input_files):
...
+                outlier_type = 'all images are outliers'
+
+            # add row of nan's if this well is an outlier
+            if well_is_outlier == 1:
+                n_stats = len(column_names) - 5
+                # FIXME in the corner case where the first well we encounter is an outlier, this will fail!
+                # @Roman, what's the best way to fix this?
+                assert n_stats > 0, "Firs well was outlier"
+                table.append([well_name, n_outlier_images, np.nan, well_is_outlier, outlier_type] + [np.nan] * n_stats)
                 continue
```
Comment (imagirom) [inline batchlib/analysis/cell_level_analysis.py:822]: I would just append `None` to the table here. Then get `n_stats` after the for loop and replace all the `None`s in the table with rows full of `np.nan`.

File: batchlib/analysis/cell_analysis_qc.py (hunk 1 above)
Comment (constantinpape) [reply batchlib/analysis/cell_analysis_qc.py:292]: Yes, I only do this for the wells. So let's keep it that way then we don't need to make it more complicated :).

File: batchlib/analysis/cell_level_analysis.py (hunk 2 above)
Comment (constantinpape) [reply batchlib/analysis/cell_level_analysis.py:691]: I just refactored all of this to the new free function `load_cell_outlier_dict`, that is also used in `WellLevelQC`. The overall functionality has stayed the same.

File: batchlib/analysis/cell_level_analysis.py (hunk 3 above)
Comment (constantinpape) [reply batchlib/analysis/cell_level_analysis.py:822]: ok, that makes a lot of sense, I will implement it.

**Comment by constantinpape** [comment]
Thanks for the comments, and no need to look at the test.

File: batchlib/analysis/cell_level_analysis.py (hunk 2 above)
Comment (imagirom) [reply batchlib/analysis/cell_level_analysis.py:691]: Ah I see, thank you for the explanation.

File: batchlib/analysis/cell_level_analysis.py (hunk 3 above)
Comment (constantinpape) [reply batchlib/analysis/cell_level_analysis.py]: Ok, I simplified it a bit more.
I am not sure if this catches all corner cases (in the case where all input images are outliers),
but I will just write a test for this and we can fix it if it's an issue.

**Comment by constantinpape** [comment]
@imagirom Ok, I went over this again and implemented a solution that works now (it's checked by one of the tests). I will wait till CI passes and then merge this.
